lib/srv/discovery/access_graph_azure.go

```diff
@@ -0,0 +1,409 @@
+			&accessgraphv1alpha.AzureEventsStreamRequest{
+				Operation: &accessgraphv1alpha.AzureEventsStreamRequest_Upsert{
+					Upsert: &accessgraphv1alpha.AzureResourceList{
+						Resources: upsert.Resources[i:end],
+					},
+				},
+			},
+		)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// azurePushDeleteInBatches deletes resources from the Access Graph in batches
+func azurePushDeleteInBatches(
+	client accessgraphv1alpha.AccessGraphService_AzureEventsStreamClient,
+	toDel *accessgraphv1alpha.AzureResourceList,
+) error {
+	for i := 0; i < len(toDel.Resources); i += batchSize {
+		end := i + batchSize
+		if end > len(toDel.Resources) {
+			end = len(toDel.Resources)
+		}
+		err := client.Send(
+			&accessgraphv1alpha.AzureEventsStreamRequest{
+				Operation: &accessgraphv1alpha.AzureEventsStreamRequest_Delete{
+					Delete: &accessgraphv1alpha.AzureResourceList{
+						Resources: toDel.Resources[i:end],
+					},
+				},
+			},
+		)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// azurePush upserts and deletes Azure resources to/from the Access Graph
+func azurePush(
+	client accessgraphv1alpha.AccessGraphService_AzureEventsStreamClient,
+	upsert *accessgraphv1alpha.AzureResourceList,
+	toDel *accessgraphv1alpha.AzureResourceList,
+) error {
+	err := azurePushUpsertInBatches(client, upsert)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = azurePushDeleteInBatches(client, toDel)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = client.Send(
+		&accessgraphv1alpha.AzureEventsStreamRequest{
+			Operation: &accessgraphv1alpha.AzureEventsStreamRequest_Sync{},
+		},
+	)
+	return trace.Wrap(err)
+}
+
+// getAllTAGSyncAzureFetchers returns both static and dynamic TAG Azure fetchers
+func (s *Server) getAllTAGSyncAzureFetchers() []*azuresync.Fetcher {
+	allFetchers := make([]*azuresync.Fetcher, 0, len(s.dynamicTAGAzureFetchers))
+
+	s.muDynamicTAGAzureFetchers.RLock()
+	for _, fetcherSet := range s.dynamicTAGAzureFetchers {
+		allFetchers = append(allFetchers, fetcherSet...)
+	}
+	s.muDynamicTAGAzureFetchers.RUnlock()
+
+	allFetchers = append(allFetchers, s.staticTAGAzureFetchers...)
+	return allFetchers
+}
+
+// initializeAndWatchAzureAccessGraph initializes and watches the TAG Azure stream
+func (s *Server) initializeAndWatchAzureAccessGraph(ctx context.Context, reloadCh chan struct{}) error {
+	// Check if the access graph is enabled
+	clusterFeatures := s.Config.ClusterFeatures()
+	policy := modules.GetProtoEntitlement(&clusterFeatures, entitlements.Policy)
+	if !clusterFeatures.AccessGraph && !policy.Enabled {
+		return trace.Wrap(errTAGFeatureNotEnabled)
+	}
+
+	// Configure the access graph semaphore for constraining multiple discovery servers
+	const (
+		semaphoreExpiration = time.Minute
+		semaphoreName       = "access_graph_azure_sync"
+		serviceConfig       = `{
+		 "loadBalancingPolicy": "round_robin",
+		 "healthCheckConfig": {
+			 "serviceName": ""
+		 }
+	 }`
```

Comment:
I'm seeing this being repeated over and over.
Can we move it to a const instead?

Reply:
we can reuse the aws one as they won't change